openvm/src/plonk/air_to_plonkish.rs

```diff
@@ -178,29 +186,50 @@ where
                     c: c.clone(),
                     ..Default::default()
                 });
-                c
+                (false, c)
             }
             AlgebraicExpression::UnaryOperation(AlgebraicUnaryOperation { op, expr }) => match op {
                 AlgebraicUnaryOperator::Minus => {
-                    let (q_o, c) = self.make_output(assert_zero);
-                    let a = self.evaluate_expression(expr, false);
-                    self.plonk_circuit.add_gate(Gate {
-                        q_l: -T::ONE,
-                        q_o,
-                        a,
-                        c: c.clone(),
-                        ..Default::default()
-                    });
-                    c
+                    let (_neg, var) = self.evaluate_expression(expr, false, false);
+
+                    if assert_zero {
+                        self.plonk_circuit.add_gate(Gate {
+                            q_l: -T::ONE,
+                            a: var.clone(),
+                            c: Variable::Unused,
+                            ..Default::default()
+                        });
+                        (true, Variable::Unused)
+                    } else if is_last {
+                        // if is_last is ture, assert_zero is false, the expression is used in bus interaction
+                        if let AlgebraicExpression::Reference(r) = &**expr {
+                            self.plonk_circuit.add_gate(Gate {
+                                q_l: -T::ONE,
+                                q_o: -T::ONE,
+                                a: Variable::Witness(r.clone()),
+                                c: Variable::Tmp(self.temp_id_offset),
+                                ..Default::default()
+                            });
+                            self.temp_id_offset += 1;
+                            (false, Variable::Tmp(self.temp_id_offset - 1))
+                        } else {
+                            self.plonk_circuit.gates.last_mut().unwrap().q_o =
+                                -T::ONE * self.plonk_circuit.gates.last().unwrap().q_o;
+                            (false, self.plonk_circuit.gates.last().unwrap().c.clone())
+                        }
+                    } else {
+                        (true, var.clone())
+                    }
```

Comment:
```suggestion
                    let (neg, var) = self.evaluate_expression(expr, false, false);
                    if is_last {
                        if neg {
                            // Just discard the inversion.
                            (false, var)
                        } else {
                            // We have to materialize the inversion.
                            let (q_o, c) = self.make_output(assert_zero);
                            self.plonk_circuit.add_gate(Gate {
                                q_l: -T::ONE,
                                q_o,
                                a: var,
                                c: c.clone(),
                                ..Default::default()
                            });
                            (false, c)
                        }
                    } else {
                        (!neg, var.clone())
                    }
```

I think this also fixes a bug, because `_neg` above is ignored, so for `-(-x)`, we would have returned `(true, x)`.